### PR TITLE
Updated Vowpal Wabbit to 7.10

### PR DESCRIPTION
**Pull request by zachmayer**
brew tests pass on my local machine.

7.8 is a big update, 7.9 and 7.10 are mostly bug fixes. Below are changelogs from VW's primary author:
Changes in vw 7.8:
1. Learning to Search: Hal completely rewrote the learning to search
system, enough that the numbers here are looking obsolete. Kai-Wei has
also created several advanced applications for entity-relation and
dependency parsing which are promising.
2. Languages Hal also created a good python library, which includes
call-backs for learning to search. You can now develop advanced
structured prediction solutions in a nice language. Jonathan Morra also
contributed an initial Java interface.
3. Exploration The contextual bandit subsystem now allows evaluation of
an arbitrary policy, and an exploration library is now factored out
into an independent library (principally by Luong with help from Sid
and Sarah). This is critical for real applications because
randomization must happen at the point of decision.
4. Reductions The learning reductions subsystem has continued to
mature, although the perfectionist in me is still dissatisfied. As a
consequence, it’s now pretty easy to program new reductions, and the
efficiency of these reductions has generally improved. Several news
ones are cooking.
5. Online Learning Alekh added an online SVM implementation based on
LaSVM. This is known to parallelize well via the para-active approach.

Changes in vw 7.9:
Primarily updates to learning reductions modularity.

Changes in vw 7.10:
No significant changes from 7.9---primarily bugfixes.

### CODE REVIEW

**Comment by mikemcquaid** [comment]
Failing on 10.8: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/21509/version=mountain_lion/console

**Comment by zachmayer** [comment]
Hmmm, I'm not sure how to fix this.  I could say it depends on 10.9 or higher, but it seems like it'd be nice to make the library work on 10.7+. The relevant error seems to be:
```
In file included from cbify.cc:6:
./../explore/static/MWTExplorer.h:16:10: fatal error: 'tuple' file not found
#include <tuple>
```

I think `<tuple>` is part of boost.  Any idea why VW could find boost on 10.10 and 10.9, but can't find it on 10.8?

**Comment by mikemcquaid** [comment]
Does it maybe need C++11? If so `ENV.cxx11` and/or `needs :cxx11` may help.

**Comment by zachmayer** [comment]
I don't think it does, but I'll add that and see what happens.

**Comment by zachmayer** [comment]
hmmm, that failed too.  Trying something else.

**Comment by zachmayer** [comment]
It's weird: mavericks and yosemite are fine.  Only mountain lion fails

**Comment by zachmayer** [comment]
Ok, now I think it IS a :cxx11 issue, but I can't seem to fix it.  One the VowpalWabbit github page, they note:
*Mac OS X 10.8 and below*: ``configure.cxx_stdlib=libc++`` and ``configure.cxx=clang++`` ensure that ``clang++`` uses
the correct C++11 functionality while building Boost. Ordinarily, ``clang++`` relies on the older GNU ``g++`` 4.2 series
header files and ``stdc++`` library; ``libc++`` is the ``clang`` replacement that provides newer C++11 functionality. If
these flags aren't present, you will likely encounter compilation errors when compiling _vowpalrabbit/cbify.cc_. These
error messages generally contain complaints about ``std::to_string`` and ``std::unique_ptr`` types missing.

To compile:
```
$ sh autogen.sh --enable-libc++
$ make
$ make test    # (optional)

File: Library/Formula/vowpal-wabbit.rb [hunk 1]
```diff
@@ -17,12 +17,72 @@ class VowpalWabbit < Formula
...
     system "make"
     system "make", "install"
   end
-end
+
+  test do
+    system bin/"vw", "--version"
+    system bin/"vw", "--help"
+
+    (testpath/"house_dataset").write("0 | price:.23 sqft:.25 age:.05 2006
```
Comment (mikemcquaid) [inline Library/Formula/vowpal-wabbit.rb]: Use can use `EOS.undent` here to nicely indent these things across lines.

File: Library/Formula/vowpal-wabbit.rb [hunk 2]
```diff
@@ -17,12 +17,72 @@ class VowpalWabbit < Formula
...
+                           "--with-boost=#{Formula['boost'].opt_prefix}",
+                           "--enable-libc++"
     system "make"
     system "make", "install"
   end
-end
+
+  test do
+    system bin/"vw", "--version"
+    system bin/"vw", "--help"
```
Comment (mikemcquaid) [inline Library/Formula/vowpal-wabbit.rb]: Don't need these two with the tests below.

File: Library/Formula/vowpal-wabbit.rb [hunk 3]
```diff
@@ -17,12 +17,72 @@ class VowpalWabbit < Formula
...
+    system bin/"vw house_dataset --constant 1"
+    system bin/"vw house_dataset --ngram 2 --skips 2"
+    system bin/"vw house_dataset --l1 0.01 --l2 0.01"
+    system bin/"vw house_dataset --loss_function quantile"
+    system bin/"vw house_dataset -l 10 -c --passes 25 --holdout_off --decay_learning_rate=.95"
+    system bin/"vw house_dataset -l 10 -c --passes 25 --holdout_off"
+    system bin/"vw house_dataset -l 10 -c --passes 25 --holdout_off -f house.model"
+    system bin/"vw house_dataset -p /dev/stdout --quiet"
+    system bin/"vw -i house.model -t house_dataset -p /dev/stdout --quiet"
+    system bin/"vw house_dataset --audit --quiet"
```
Comment (mikemcquaid) [inline Library/Formula/vowpal-wabbit.rb]: Probably only need a single `vw` invocation on some data and `"separate", "strings", "like", "this"`

**Comment by tdsmith** [comment]
boost looks like a runtime dependency to me:
```
/usr/local/Cellar/vowpal-wabbit/7.7/lib/liballreduce.0.dylib:
	/usr/local/lib/liballreduce.0.dylib (compatibility version 1.0.0, current version 1.0.0)
	/usr/local/lib/libboost_program_options-mt.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.5)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
```

**Comment by zachmayer** [comment]
@tdsmith I just added a commit that makes boost a runtime requirement.  We'll see what happens.

The interesting thing is, I've had sever build succeed for mavericks and yosemite with `depends_on "boost" => :build`

**Comment by zachmayer** [comment]
@tdsmith Still failing on mountain lion: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/21611/version=mountain_lion/testReport/junit/brew-test-bot/mountain_lion/test_vowpal_wabbit/

**Comment by zachmayer** [comment]
Hey!  It worked!

**Comment by zachmayer** [comment]
I'm going to try taking stuff out until I have the minimal formula that will compile on all 3 platforms.

Once I'm done, should I squash all these commits into a single commit?  Sorry it got a little out of control-- I don't have a 10.8 machine to test on.

**Comment by mikemcquaid** [comment]
> Once I'm done, should I squash all these commits into a single commit? Sorry it got a little out of control-- I don't have a 10.8 machine to test on.

Yep, if you don't mind, thanks!

**Comment by zachmayer** [comment]
Hmm, so the latest commit failed, but the jenkins log shows everything passing.

**Comment by tdsmith** [comment]
Please don't mark boost build-only.

So people don't have to rebuild boost when they don't need to, you can write something like:

```ruby
if MacOS.version < :mavericks
  depends_on "boost" => "c++11"
else
  depends_on "boost"
end
```

**Comment by zachmayer** [comment]
Is it usual to wait this long for Jenkins to run?  It's been like 12 hours or so.

**Comment by mikemcquaid** [comment]
@BrewTestBot test this please

**Comment by zachmayer** [comment]
Ok I squashed all the commits together, and I think this builds on all 3 platforms (but we'll see what Jenkins says).  @mikemcquaid @tdsmith How does this look now?  Anything else that needs work?

Thanks!

**Comment by zachmayer** [comment]
@BrewTestBot test this please

**Comment by DomT4** [comment]
> test this please

Afraid asking the test bot to do things is limited to the maintainers, but you'll have triggered a new build by squashing the commits together anyway :).

**Comment by zachmayer** [comment]
hahaha, thanks for the tip.  This is my first brew PR =D

**Comment by DomT4** [comment]
No worries, I did the same thing in quite a few of my PRs until I finally asked the maintainers why the bot was just ignoring the heck out of me :smile_cat:. Nice! Always cool to see more people contributing, Welcome! The bot is green on all three platforms :).

**Comment by zachmayer** [comment]
:metal:

**Comment by tdsmith** [comment]
Awesome, looks great. For future reference, we prefer commit titles like "vowpal-wabbit 7.10" for version bumps. Thanks a lot for submitting a PR and for your patience with revisions; Homebrew depends on contributions like these!

**Comment by zachmayer** [comment]
@tdsmith I'll keep that in mind in the future.  Thanks for helping me learn the system!

**Pull request closed without merging**